cardano-node/src/Cardano/CLI/Delegation/PoolRegistry.hs

```diff
@@ -0,0 +1,301 @@
+
+  let (id, ext) = Path.splitExtension file
+      sigPath = dir </> id <.> "sig"
+
+  unlessM (lift $ Dir.doesFileExist sigPath) $
+    checks ["Missing signature file: " <> pack sigPath]
+  unlessM (lift $ Dir.readable <$> Dir.getPermissions sigPath) $
+    checks ["Signature file not readable: " <> pack sigPath]
+  checks $
+    ["Submission filename doesn't have .json extension: " <> pack file
+    | ext /= ".json" ]
+
+  pubkey <- case validatePublicKey (pack id) of
+    Left e -> err $
+      ["Submission filename (actually public key) invalid: " <> e]
+    Right x -> pure x
+
+  contents <- lift $ BSL.fromStrict <$> BS.readFile afp
+  case AE.eitherDecode contents of
+    Left e -> err $ ["Submission content:\n" <> pack e]
+    Right entry -> do
+
+      checks $
+        [ pack $ "Internal 'id' doesn't match filename: "
+          <> show (unId $ sId entry)  <> " vs. " <> show (unId pubkey)
+        | sId entry /= pubkey ]
+
+      -- All cheap checks done, let's do the expensive one now.
+      dupTickers <- withExceptT (PoolRegistryValidationErrors fp . (:[]) .
+                               ("While reading all registry entries: "<>)) $
+        duplicateTickers dir
+
+      checks $
+        [ "Duplicate tickers: " <>
+          pack (unpack $ Text.intercalate ", " $ unTicker <$> dupTickers)
+        | not (null dupTickers) ]
+
+      -- All valid.
+      pure entry
+
+ where
+   checks :: [Text] -> ExceptT CliError IO ()
+   checks [] = ExceptT . pure . Right $ ()
+   checks es = err es
+
+   err :: [Text] -> ExceptT CliError IO a
+   err es = ExceptT . pure . Left  $ PoolRegistryValidationErrors fp es
+
+   duplicateTickers :: FilePath -> ExceptT Text IO [Ticker]
+   duplicateTickers dir = do
+     paths <- ((dir </>) <$>) <$> (lift $ Dir.listDirectory dir)
+     -- We maintain the invariant that 'registry' contains only files.
+     tickers <- withExceptT pack $ ExceptT $
+       sequence <$> traverse entryTicker (selectEntries paths)
+
+     pure . Set.toList $ duplicates tickers
+     where
+       selectEntries :: [FilePath] -> [FilePath]
+       selectEntries paths =
+         flip filter paths $ (==".json") . Path.takeExtension
+
+       entryTicker :: FilePath -> IO (Either String Ticker)
+       entryTicker f =
+         (bimap (("While parsing '"<>f<>"': ")<>) sTicker) . AE.eitherDecode . BSL.fromStrict
+           <$> BS.readFile f
+
+--------------------------------------------------------------------------------
+-- * Ancillary
+--
+-- | Submission identifier is actually a 32-byte base16-encoded public key.
+--   We additionally require that the hexadecimal must be all-lowercase.
+validatePublicKey :: Text -> Either Text Id
+validatePublicKey text =
+  if      not (length s == 64 && all Char.isHexDigit s)
+  then fail $ "Public key not a 64-char-long base16: " <> take 128 s
+  else if not ((Char.toLower <$> s) == s)
+  then fail $ "Public key not all-lowercase: " <> s
+  else Right $ Id text
+  where s = unpack text
```

Comment:
I'd suggest using [convertFromBase](https://hackage.haskell.org/package/memory-0.15.0/docs/Data-ByteArray-Encoding.html#v:convertFromBase):

```hs
convertPublicKey :: Text -> Either String Id
convertPublicKey txt = do
    bytes <- convertFromBase Base16 (T.encodeUtf8 txt)
    when (BS.length bytes /= 32) $ fail ...
    ...
```

Reply:
I'd say we want to be stricter than that -- we want to enforce that all:

1. filenames, and
2. `id` fields are lowercase

That simply avoids uniqueness issues.